File: app/src/main/java/com/eighteengray/procamera/MainActivity.java
package com.eighteengray.procamera;

import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Debug;
import android.view.WindowManager;
import com.eighteengray.procamera.camera.Camera2Fragment;
import com.eighteengray.procamera.camera.RecordVideoFragment;
import com.eighteengray.procameralibrary.common.Constants;
import com.eighteengray.procameralibrary.dataevent.ModeSelectEvent;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;


public class MainActivity extends AppCompatActivity {
    Camera2Fragment camera2Fragment;
    RecordVideoFragment recordVideoFragment;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Debug.startMethodTracing("ProCamera");
        //去掉status bar
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
        setContentView(R.layout.activity_main);

        /*StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                .detectAll()//开启所有的detectXX系列方法
                .penaltyDialog()//弹出违规提示框
                .penaltyLog()//在Logcat中打印违规日志
                .build());

        StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                .detectActivityLeaks()//检测Activity泄露
                .penaltyLog()//在Logcat中打印违规日志
                .build());*/
        EventBus.getDefault().register(this);
        initFragment();
    }


    private void initFragment() {
        camera2Fragment = new Camera2Fragment();
        recordVideoFragment = new RecordVideoFragment();
        replaceFragment(camera2Fragment);
    }


    private void replaceFragment(Fragment fragment) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.rl_aty_main, fragment);
        fragmentTransaction.commit();
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
        Debug.stopMethodTracing();
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onModeSwitch(ModeSelectEvent modeSelectEvent) {
        switch (modeSelectEvent.getMode()) {
            case Constants.MODE_CAMERA:
                replaceFragment(camera2Fragment);
                break;

            case Constants.MODE_RECORD:
                replaceFragment(recordVideoFragment);
                break;
        }
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        camera2Fragment.actionBarDrawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        camera2Fragment.actionBarDrawerToggle.onConfigurationChanged(newConfig);
    }

}

File: app/src/main/java/com/eighteengray/procamera/MainActivity.kt
package com.eighteengray.procamera

import android.content.pm.ActivityInfo
import android.os.Bundle
import android.view.WindowManager
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.eighteengray.procamera.camera.Camera2Fragment
import com.eighteengray.procamera.camera.RecordVideoFragment
import com.eighteengray.procameralibrary.common.Constants
import com.eighteengray.procameralibrary.dataevent.ModeSelectEvent
import com.yanzhenjie.permission.AndPermission
import com.yanzhenjie.permission.runtime.Permission
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode


class MainActivity: FragmentActivity() {
    private var camera2Fragment: Camera2Fragment? = null
    private var recordVideoFragment: RecordVideoFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //去掉status bar
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR
        setContentView(R.layout.activity_main)

        EventBus.getDefault().register(this)
        initFragment()
    }

    private fun initFragment() {
        AndPermission.with(this)
            .runtime()
            .permission(Permission.CAMERA, Permission.WRITE_EXTERNAL_STORAGE)
            .onGranted {
                camera2Fragment = Camera2Fragment()
                recordVideoFragment = RecordVideoFragment()
                replaceFragment(camera2Fragment)
            }
            .onDenied {

            }
            .start()
    }

    private fun replaceFragment(fragment: Fragment?) {
        if(fragment != null) {
            val fragmentTransaction = supportFragmentManager.beginTransaction()
            fragmentTransaction.replace(R.id.frame_layout, fragment)
            fragmentTransaction.commit()
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onModeSwitch(modeSelectEvent: ModeSelectEvent) {
        when (modeSelectEvent.mode) {
            Constants.MODE_CAMERA -> replaceFragment(camera2Fragment)
            Constants.MODE_RECORD -> replaceFragment(recordVideoFragment)
        }
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }

}